### ISSUE TITLE
optirun takes 100% CPU usage

### ISSUE THREAD

**Issue by bdbai** [Bumblebee-Project/Bumblebee#911]
Hi there,

I've got a laptop with GTX1050 and Intel graphic card. After installing bumblebee on openSUSE, everything works fine except that when I run `optirun glxgears`, one of the CPUs starts 100% working. Simply `glxgears` is just fine. 

What else should I provide to fix it?

**Comment by 0x041E** [comment]
optirun with virtualgl backend by default runs without vsync pushing out as many frames as possible therefore using also as much CPU as possible and it's equivalent to running glxgears without vsync on the intel gpu with `vblank_mode=0 glxgears`. Especially at graphically unintensive program as glxgears where GPU pushes out thousands of frames and CPU has to transport all the frames it's natural you will have high CPU usage. If you run just `glxgears` on intel gpu it's usually vsynced by default pushing out only limited(screen refresh rate) amount of frames naturally having lower CPU usage. You might want to take a look at primus if your distribution provides it and then using `primusrun` instead of `optirun`. Primus is different backend for bumblebee which should have lower CPU overhead and it by default push out maximum as many frames as your screen refresh rate therefore not using as much cpu as possible.

**Comment by bdbai** [comment]
@0x041E Thanks for your explanation! I tried `optirun -b primus glxgears` and CPU usage stayed at a low level. Maybe I should only run GPU intensive programs with optirun/VirtualGL. Is it correct?

**Comment by 0x041E** [comment]
No running GPU intensive programs with `optirun -b primus` or `primusrun` is completely fine it only limit FPS to maximum of your monitor refresh rate by default which is for power management purpose to not render frames you wouldn't see anyway(eg. 100fps on 60hz monitor). In case you just want maximum amount of frames while using primus you can prepend the command with vblank_mode=0 such as `vblank_mode=0 optirun -b primus glxgears` or `vblank_mode=0 primusrun glxgears` which turns off the vsync allowing primus to render as many frames as it can. 
To my knowledge(i'm not developer of bumblebee) the difference between VirtualGL and Primus is that Primus is client-side only, allowing it to render and display frames only on the current machine which allows it to transport the frames with lower overhead. VirtualGL on the other hand is implementation that allows the frames to even be transfered between X servers even over network therefore it uses server which has higher overhead because it transfers frames in different formats jpeg,rgb,xv,yuv,proxy as opposed to Primus which transfers frames as buffer objects by default.

**Comment by bdbai** [comment]
@0x041E it's said that primus is not as efficient as VirtualGL and the problem is not with frame rate. But I haven't tested to see which bridge is more efficient. From now on I may have to stick to primus because otherwise my CPU fan would make unbearable noise :)

**Comment by Thulinma** [comment]
For reference: primus is better than VirtualGL in every way.
Some people think it's less efficient because they see lower frame rates in certain applications, but it is actually more efficient. The frame rate is no good measurement of speed, especially in benchmarking applications such as glxgears, since they are a relatively simple operation repeated many times.
There is no reason to still be using VirtualGL unless your application or system doesn't work with primus somehow. If it works, it is better.

That said, for some applications the iGPU (e.g. Intel) is a better option than using the dGPU (e.g. Nvidia).
This is because there will always be some overhead to copy the frames (regardless of method used), and for simple scenes the iGPU will be faster than the dGPU+overhead. The more complicated the scene is, the more gain you will see from switching to the dGPU.

TLDR: Use iGPU (no bumblebee at all) for simple applications, use primus for complicated applications, use VirtualGL if primus doesn't work.

**Comment by bdbai** [comment]
Thanks @0x041E and @Thulinma ! I decide to use primus as the default bridge.

**Comment by Lekensteyn** [comment]
Yay, case solved
